feature/gnoi/os/tests/osinstall/osinstall_test.go

```diff
@@ -270,15 +270,28 @@ func (tc *testCase) verifyInstall(ctx context.Context, t *testing.T) {
 	}
 
 	if tc.dualSup {
-		if got, want := r.GetVerifyStandby().GetVerifyResponse().GetActivationFailMessage(), ""; got != want {
-			t.Errorf("OS.Verify Standby ActivationFailMessage: got %q, want %q", got, want)
-		}
+		var counter int
+		const maxRetries = 15
+		for i := 1; i < maxRetries; i++ {
+			r, err := tc.osc.Verify(ctx, &ospb.VerifyRequest{})
```

Comment:
`r, err := tc.osc.Verify(ctx, &ospb.VerifyRequest{})` becomes a duplicated code as above. maybe consider refactoring it to a quick helper func.

Reply:
if we move at end of function it will consider `r` declared but not used. I think it's best we keep it at the top of loop so for every run it's updated for evaluation

Reply:
Okay, I just realized what this is doing:

 1. Lines 260-270 (before the loop started), the code checks only the primary supervisor.
 2. The retry loop (guarded by whether the device has dual supervisors) only checks the standby supervisor.
 
So the two code sections serve a different purpose.

Clarification question: why do we retry only for the standby and not primary? Should we just retry for both?

```
func (tc *testCase) verifyInstall(...) {
  for ... {
    r, err := tc.osc.Verify(...)
    ... // check err
    ok := true
    ... // check primary supervisor and set ok false if got != want
    if tc.dualsup {
      ... // check backup supervisor and set ok false if got != want
    }
    if ok {
      t.Log("OS.Verify complete")
      return
    }
    if i == maxAttempt {
      t.Fatalf(...)
    }
    t.Log("Attempt %d of %d failed. Sleep and then retrying...", i, maxRetries)
    time.Sleep(...)
  }
}
```

Reply:
@liulk my reading of the code is that:
* As for the RPC call, there is no diff between primary and secondary status. It's the same `r, err := tc.osc.Verify(ctx, &ospb.VerifyRequest{})` that returns a struct `VerifyResponse` which contains both the primary and secondary info. 
* `if tc.dualSup {` just does an extra read about `VerifyStandby` in the `VerifyResponse` struct. 
* That's why I was saying that the retry of the RPC `r, err := tc.osc.Verify(ctx, &ospb.VerifyRequest{})` should happen at the end of the for loop, to save one extra involking (because it's already called above). 

@Hobbydos, regarding
```
I think it's best we keep it at the top of loop so for every run it's updated for evaluation
```
* If you keep it at the bottom, it still does the same refresh, right? And you save one extra duplicated invoking of the RPC.

```
if we move at end of function it will consider r declared but not used. 
```
Because you are "refreshing" the variable `r`, you don't need the colon to initial them as new variables, right? `r, err = tc.osc.Verify(ctx, &ospb.VerifyRequest{})` 

Also, @Hobbydos we prefer not to close a comment (e.g. this one) that's not yet actually closed. There were still question from @liulk and I haven't got a chance to reply your comment :).Thanks.

Reply:
@xw-g did you mean something like this instead?

```go
	if tc.dualSup {
		const maxRetries = 15
		for i := 1; i <= maxRetries; i++ {
			t.Logf("Attempt %d of %d", i, maxRetries)
			if err != nil {
				t.Fatalf("OS.Verify request failed: %s", err)
			}
			if got, want := r.GetVerifyStandby().GetVerifyResponse().GetActivationFailMessage(), ""; got != want {
				t.Errorf("OS.Verify Standby ActivationFailMessage: got %q, want %q", got, want)
			}
			if got, want := r.GetVerifyStandby().GetVerifyResponse().GetVersion(), *osVersion; got != want {
				t.Logf("OS.Verify Standby Version: got %q, want %q", got, want)
			} else {
				break
			}
			if i == maxRetries {
				t.Fatalf("OS.Verify Standby could not obtain version status in %d retries", maxRetries)
			}
			time.Sleep(1 * time.Minute)
			r, err = tc.osc.Verify(ctx, &ospb.VerifyRequest{})
		}
	}
	t.Log("OS.Verify complete")
}
```

Reply:
I just realized that if we're retrying here, potentially we don't have to wait for reboot to complete (lines 161-187). This way, the loop here will wait for both the primary and backup supervisors to report the correct version. And instead of maxRetries, it should honor *timeout.

@bstoll what do you think?

Reply:
@Hobbydos yes, that's what I meant.

@liulk I think line 161-187 is for waiting on reboot, but this loop here is waiting on the sync between dual Supervisors (after reboot). For some OS implementations, until the sync is done the RPC call won't reply with the version info etc. Therefore, I think we still need both.

Reply:
We do have to make a small modification for the verify loop so it will retry on `status.Code(err) == codes.Unavailable` rather than t.Fatal. After this, the reboot loop is redundant because the verify loop will now correctly wait for the reboot to finish in addition to checking the OS versions.

Reply:
Your conditional will mistakenly retry if err is nil. This is why your test will eventually time out and fatal when the operations were successful. We also want to retry for both single and dual supervisor devices, not just the dual.

Can you try my rewrite of verifyInstall and let me know if the code makes sense to you?

```
func (tc *testCase) verifyInstall(ctx context.Context, t *testing.T) {
	deadline := time.Now().Add(*timeout)
	for time.Now().Before(deadline) {
		r, err := tc.osc.Verify(ctx, &ospb.VerifyRequest{})
		switch status.Code(err) {
		case codes.OK:
		case codes.Unavailable:
			t.Log("Reboot in progress.")
			time.Sleep(time.Minute)
			continue
		default:
			t.Fatalf("OS.Verify request failed: %v", err)
		}

		if got, want := r.GetActivationFailMessage(), ""; got != want {
			t.Fatalf("OS.Verify ActivationFailMessage: got %q, want %q", got, want)
		}
		if got, want := r.GetVersion(), *osVersion; got != want {
			t.Log("Reboot not started yet.")
			time.Sleep(time.Minute)
			continue
		}

		if tc.dualSup {
			if got, want := r.GetVerifyStandby().GetVerifyResponse().GetActivationFailMessage(), ""; got != want {
				t.Fatalf("OS.Verify Standby ActivationFailMessage: got %q, want %q", got, want)
			}

			if got, want := r.GetVerifyStandby().GetVerifyResponse().GetVersion(), *osVersion; got != want {
				t.Log("Standby not ready.")
				time.Sleep(time.Minute)
				continue
			}
		}

		t.Log("OS.Verify complete")
		return
	}

	t.Fatal("OS.Verify did not return the correct version before deadline.")
}
```

Reply:
With the re-works of the `rebootDUT()` function, your code is more holistic than my initial patch-attempt to simply loop over if it's `dualSup`. I think having timeout` instead of `maxRetries` is better - I should've done this from the get-go but was sidelined with the initial constraint of looping through for 15 minutes. 


I tested it in my lab, the execution time remains similar and the test passed for a DualSup DUT

```
    osinstall_test.go:291: OS.Verify complete
--- PASS: TestOSInstall (1100.99s)
PASS

*** Releasing the testbed...

ok      command-line-arguments  1118.730s
```

Reply:
Update the PR with my proposal?

Reply:
I'll have a commit pushed shortly.